File: strings/longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
function lengthOfLongestSubstring(s) {
  const n = s.length;
  let maxLength = 0;
  const indexMap = {};

  let j = 0;

  for (let i = 0; i < n; i++) {
    const curr = s[i];
    if (indexMap[curr]) {
      j = Math.max(indexMap[curr], j);
    }

    const currLength = i - j + 1;

    maxLength = Math.max(currLength, maxLength);

    indexMap[curr] = i + 1;
  }
  return maxLength;
}

File: strings/longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.go
package longestsubstringwithoutrepeatingcharacters

func lengthOfLongestSubstring(s string) int {
	n := len(s)
	maxLength := 0
	indexMap := make(map[byte]int)

	j := 0

	for i := 0; i < n; i++ {
		curr := s[i]

		if indexMap[curr] != 0 {
			if indexMap[curr] > j {
				j = indexMap[curr]
			}
		}

		currLength := i - j + 1

		if currLength > maxLength {
			maxLength = currLength
		}

		indexMap[curr] = i + 1

	}

	return maxLength

}
